### PR TITLE
CON-2259 Validate the delivery status of a notification

### PR DESCRIPTION
**Pull request by VasanthaKasirajan3008**

### CODE REVIEW

File: src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs [hunk 1]
```diff
@@ -22,6 +22,9 @@ public class DeliveryNotificationCommand : IDeliveryNotificationCommand
         private readonly ICertificateService _certificateService;
         private readonly IFileTransferClient _fileTransferClient;
         private readonly SftpSettings _sftpSettings;
+        public const string Delivered = "Delivered";
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs:25]: There is an Assessor API client library in the project - could these go there so they could be reused as they belong to Assessor API really.

Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs]: done

File: src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs [hunk 2]
```diff
@@ -15,5 +15,7 @@ public static class CertificateStatus
         public const string Deleted = "Deleted";
         public const string Reprint = "Reprint";
         public const string Error = "Error";
+        public const string Delivered = "Delivered";
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs:18]: If you look at CertificateStatus in both Assessor and Admin - they are currently the same duplicated code - this should be another duplicate of it - and you'll set that in Assessor and Admin it has the style of sets of status's as well which you can use

File: src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs [hunk 3]
```diff
@@ -21,7 +22,8 @@ public class DeliveryNotificationCommand : IDeliveryNotificationCommand
         private readonly ILogger<DeliveryNotificationCommand> _logger;
         private readonly ICertificateService _certificateService;
         private readonly IFileTransferClient _fileTransferClient;
-        private readonly SftpSettings _sftpSettings;
+        private readonly SftpSettings _sftpSettings;       
+        public static string[] DeliveryNotificationStatus = new[] { DeliveryNotification.CertificateStatus.Delivered, DeliveryNotification.CertificateStatus.NotDelivered };
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs:26]: See comment about CertificateStatus above

File: src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs [hunk 4]
```diff
@@ -69,6 +71,19 @@ private async Task ProcessFile(string fileName)
                 return;
             }
 
+            var invalidDeliveryNotificationStatuses = receipt.DeliveryNotifications
+                   .GroupBy(certificateDeliveryNotificationStatus => certificateDeliveryNotificationStatus.Status)
+                   .Select(certificateDeliveryNotificationStatus => certificateDeliveryNotificationStatus.Key)
+                   .Where(deliveryNotificationStatus => !DeliveryNotificationStatus.Contains(deliveryNotificationStatus))
+                   .ToList();
+
+            //Scenario 4:  Invalid status in delivery notification file
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs:80]: Comments like this don't actually refer to anything useful to reword them or remove them

**Review by ChrisJWoodcock — COMMENTED** [review]
Rename your PR please

File: src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs (hunk 4 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.Assessor.Functions/Domain/Print/DeliveryNotificationCommand.cs]: removed

File: src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs (hunk 2 above)
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs]: Copied from Assessor service

File: src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs [hunk 5]
```diff
@@ -1,4 +1,6 @@
-﻿namespace SFA.DAS.Assessor.Functions.ExternalApis.Assessor.Constants
+﻿using System.Linq;
+
```
Comment (ChrisJWoodcock) [inline src/SFA.DAS.Assessor.Functions.ExternalApis/Assessor/Constants/CertificateStatus.cs:2]: Apologies for this..... now that you have copied this across I can see how wrong that is - you were correct originally the definition of 'Delivered' or 'NotDelivered' is a string which belongs in the print command it should not be tied to the  CertificateStatus - which is really a co-incidence that the strings are the same.

After putting it back - could you Delete the CertificateStatus from the function app - as it really should not be there

**Review by ChrisJWoodcock — CHANGES_REQUESTED** [review]

**Review by ChrisJWoodcock — APPROVED** [review]

**Pull request merged**